### PR TITLE
Use logical instead of placed volumes for VolumeId

### PR DESCRIPTION
**Pull request by mrguilima**
As discussed in today's (May 6) physics hackathon, the GeoParams and GeoTrackView interfaces are being modified to return logical volumes rather than physical volumes.  For example:
- GeoParams::num_volumes() now returns the number of logical volumes 
- GeoParams::id_to_label() and label_to_id() now return the label or id of that logical volume
- GeoTrackView::volume() and volume_id() now return the logical volume (and its ID) containing the current track

Geometry tests were adapted to the new volume IDs.
*Note:* updated version of VecGeom required - https://gitlab.cern.ch/VecGeom/VecGeom/-/merge_requests/816

### CODE REVIEW

**Review by sethrj — APPROVED** [review]
Looks great! Does it need a new vecgeom tag?

**Comment by mrguilima** [comment]
Oh yes, I forgot to mention it in the PR description!

File: src/geometry/GeoParams.cc
Comment (stognini) [inline src/geometry/GeoParams.cc:110]: The `vecgeom::LogicalVolume` has a `std::string GetLabel()` and a `const char* GetName()`. Do you know if there is any particular reason for having two different functions, with different names, that return essentially the same information? Especially if the terms `label` and `name` could have different meanings in vecgeom -- which I don't know if it is the case, but it is worth pointing that out.

**Comment by sethrj** [comment]
Okeydoke. It  looks like [the CI is also failing](https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Celeritas/detail/PR-232/1/pipeline) because the logical volume id is an `int` whereas the physical volume is `unsigned int`. If we have to update the vecgeom version, can you also update the GeoParams constructor to use the new "num logical volumes" method that you added rather than the vector size workaround the @stognini devised? Once the new vecgeom gets merged (and perhaps a new tag could be created soon?) then I'll update the CI on this branch and push a new docker image.

File: test/geometry/LinearPropagator.test.cc
Comment (stognini) [inline test/geometry/LinearPropagator.test.cc:99]: The comments in lines `79` and `159` state that `VolumeId{2}` represents `Shape1 -> Envelope`, while the comment here says `World -> Envelope`.  Could you clarify this?

Comment (sethrj) [reply test/geometry/LinearPropagator.test.cc:99]: Even better, since this is a higher-level test than the geometry itself, it would be better for readability and maintainability  to directly compare the expected volume names rather than the raw volume ids:
```suggestion
        EXPECT_EQ("Envelope", geom.id_to_label(step.volume));
```

File: src/geometry/GeoParams.cc
Comment (mrguilima) [reply src/geometry/GeoParams.cc:110]: These different terms, label and name, are basically interchangeable in VecGeom.  I think the original was label, and then the name was added for easy compatibility with Geant4.

**Review by stognini — APPROVED** [review]

File: test/geometry/LinearPropagator.test.cc
Comment (mrguilima) [reply test/geometry/LinearPropagator.test.cc:99]: > Even better, since this is a higher-level test than the geometry itself, it would be better for readability and maintainability to directly compare the expected volume names rather than the raw volume ids:

Thanks, I like this a lot!  Implementing...

**Comment by sethrj** [comment]
OK @mrguilima is the change of the "logical" id from int to unsigned int happening in the upstream vecgeom code, or in a commit to this branch? I think consistency on the vecgeom side would be better than casting on our side.

File: test/geometry/LinearPropagator.test.cc
Comment (mrguilima) [reply test/geometry/LinearPropagator.test.cc:99]: > The comments in lines `79` and `159` state that `VolumeId{2}` represents `Shape1 -> Envelope`, while the comment here says `World -> Envelope`. Could you clarify this?

`VolumeId{2}` corresponds to the "Envelope" volume, which is the name of the "post-step" volume in both cases.

**Comment by mrguilima** [comment]
I have updated the geometry tests, but in order to pass the CI tests, the current master of VecGeom (e.g. commit 0c163e1) is required.  That's the reason for the current failure in CI testing.

**Comment by mrguilima** [comment]
> OK @mrguilima is the change of the "logical" id from int to unsigned int happening in the upstream vecgeom code, or in a commit to this branch? I think consistency on the vecgeom side would be better than casting on our side.

I tried to change `VecGeom LogicalVolume::id()` return type from `int` to `unsigned int`.  Andrei said that would not be ok, because in the GPU this function returns -1 to flag an invalid volume ID.  So that function still returns `int`, and I had to cast it to `unsigned int` on the celeritas side.  It remains to be seen whether the -1 returned from VecGeom would create problems, as I suppose an invalid volume would never be returned inside the world volume.

**Comment by stognini** [comment]
>  Andrei said that would not be ok, because in the GPU this function returns -1 to flag an invalid volume ID

Is there any reason to return `-1`? Seems like an invalid volume might be enough of a problem to return an assertion failure and stop the execution -- which would resolve the `unsigned int` / `int` incompatibility

**Comment by sethrj** [comment]
@mrguilima How's this coming along? When you create PRs on celeritas, can you cross-link them to the prerequisite VecGeom pull requests so I can check their progress and make sure they're merged and tagged before trying to update the CI and dev systems? Thanks!

**Comment by mrguilima** [comment]
Hi Seth,
I had some build issues, which resulted in a VecGeom CMake config fix by Ben Morgan committed yesterday. I was waiting for his commit to be merged, but it is not really mandatory.
So I have just rebased it and renewed the merge request (https://gitlab.cern.ch/VecGeom/VecGeom/-/merge_requests/816).
Now, is this link good enough for the cross-link you requested?  If not, please let me know how to cross-link.

**Comment by sethrj** [comment]
Great! That's just what I need. If 1.1.15 is coming up soon let's wait for that (ideal); otherwise I can specify a git hash after your latest PR (less than ideal).

**Comment by mrguilima** [comment]
Andrei has merged the required VecGeom commit.
Now, considering the ongoing vCHEP activities, I am not sure when the next VecGeom tag will get defined.

**Pull request merged**
